src/atomate2/vasp/flows/defect.py

```diff
@@ -22,28 +30,127 @@
 DEFECT_INCAR_SETTINGS = {
     "ISMEAR": 0,
-    "LWAVE": True,
     "SIGMA": 0.05,
     "KSPACING": None,
     "ENCUT": 500,
 }
-DEFECT_KPOINT_SETTINGS = {"reciprocal_density": 64}
+DEFECT_KPOINT_SETTINGS = Kpoints(
+    comment="special k-point",
+    num_kpts=1,
+    style=Kpoints_supported_modes.Reciprocal,
+    kpts=((0.25, 0.25, 0.25),),
+    kpts_shift=(0, 0, 0),
+    kpts_weights=[1],
+)
 
-DEFECT_RELAX_GENERATOR = AtomicRelaxSetGenerator(
+DEFECT_RELAX_GENERATOR: AtomicRelaxSetGenerator = AtomicRelaxSetGenerator(
     use_structure_charge=True,
     user_incar_settings=DEFECT_INCAR_SETTINGS,
     user_kpoints_settings=DEFECT_KPOINT_SETTINGS,
 )
-DEFECT_STATIC_GENERATOR = StaticSetGenerator(
+
+DEFECT_STATIC_GENERATOR: StaticSetGenerator = StaticSetGenerator(
     user_incar_settings=DEFECT_INCAR_SETTINGS,
     user_kpoints_settings=DEFECT_KPOINT_SETTINGS,
 )
 
 
+################################################################################
+# Formation Energy                                                            ##
+################################################################################
+
+
+@dataclass
+class FormationEnergyMaker(Maker):
+    """Class to generate VASP input sets for the calculation of the formation energy
+    diagram.
+
+    Parameters
+    ----------
+    name: str
+        The name of the flow created by this maker.
+    relax_maker: .BaseVaspMaker or None
+        A maker to perform a atomic-position-only relaxation on the defect charge
+        states.
+        If None, the defaults will be used.
+    """
+
+    name: str = "formation energy"
+
+    relax_maker: BaseVaspMaker = field(
+        default_factory=lambda: RelaxMaker(
+            input_set_generator=DEFECT_RELAX_GENERATOR,
+            task_document_kwargs={"store_volumetric_data": ["locpot"]},
+        )
+    )
+
+    def __post_init__(self):
+        """Post-initialization."""
+        self.relax_maker = self.relax_maker.update_kwargs(
+            update={
+                "_set": {
+                    "input_set_generator->user_incar_settings->LVHAR": True,
+                    "input_set_generator->use_structure_charge": True,
+                }
+            },
+            class_filter=RelaxMaker,
+            dict_mod=True,
+        )
+        self.relax_maker = self.relax_maker.update_kwargs(
+            update={
+                "_add_to_set": {"task_document_kwargs->store_volumetric_data": "locpot"}
+            },
+            class_filter=RelaxMaker,
+            dict_mod=True,
+        )
+
+    def make(
+        self,
+        defect_gen: DefectGenerator,
+        dielectric: float | NDArray,
+        bulk_sc_dir: str | Path | None = None,
+        sc_mat: NDArray | None = None,
+    ):
+        """Make a flow to calculate the formation energy diagram."""
+        jobs = []
+
+        if bulk_sc_dir is None:
+            get_sc_job = bulk_supercell_calculation(
```

Comment:
Currently this logic is quite hard to follow as `bulk_supercell_calculation` and `get_supercell_from_prv_calc` do several things.

Could you break it up into:
1. Getting the supercell matrix.
2. Getting the bulk_supercell_dir.

Basically the logic would then be very explicit and look like this. Before running everything else:

If bulk_dir provided:
- Add job to extract supercell (and check supercell matches provided supercell if not None).
- Set supercell_mat = supercell_extract.output
- Leave bulk_dir as is.

If bulk not provided:
- Add job to calculate supercell matrix.
- Add job to calculate bulk
- Set supercell_mat = get_supercell.output
- Set bulk_dir = bulk_job.output.dir_name

Afterwards use same logic to submit jobs.

That way, if we want to reuse the code to generate the supercell matrix that will be easy to do. At the moment the supercell logic is tied up with generating a bulk calculation.